presto-array/src/main/java/com/facebook/presto/array/ReferenceCountMap.java

```diff
@@ -14,39 +14,42 @@
 package com.facebook.presto.array;
 
 import io.airlift.slice.SizeOf;
-import it.unimi.dsi.fastutil.objects.Object2IntOpenCustomHashMap;
+import it.unimi.dsi.fastutil.longs.Long2IntOpenHashMap;
 import org.openjdk.jol.info.ClassLayout;
 
+/**
+ * This class tracks how many time an objects have been referenced
+ * in order not to over count memory in complex objects such as SliceBigArray or BlockBigArray.
+ * Ideally the key set should be the object references themselves.
+ * But it turns out such implementation can lead to GC performance issues.
+ * With several attempts including breaking down the object arrays into object big arrays,
+ * the problem can be solved but tracking overhead is still high.
+ * Benchmark results show using hash maps with primitive arrays perform way better than object arrays.
+ * Therefore, we use identity hash code + size of an object to identify an object.
```

Comment:
This explanation is good, but is slightly unsuitable for javadoc. Maybe have it in normal comment, perhaps accompanying actual code.